sync-core/src/main/java/com/cloudant/sync/query/UnindexedMatcher.java

```diff
@@ -98,14 +106,221 @@ public static UnindexedMatcher matcherWithSelector(Map<String, Object> selector)
         return matcher;
     }
 
+    @SuppressWarnings("unchecked")
     private static ChildrenQueryNode buildExecutionTreeForSelector(Map<String, Object> selector) {
         // At this point we will have a root compound predicate, AND or OR, and
         // the query will be reduced to a single entry:
         // { "$and": [ ... predicates (possibly compound) ... ] }
         // { "$or": [ ... predicates (possibly compound) ... ] }
 
-        // TODO - build execution tree for selector
-        return null;
+        ChildrenQueryNode root = null;
+        List<Object> clauses = new ArrayList<Object>();
+
+        if (selector.get(AND) != null) {
+            clauses = (List<Object>) selector.get(AND);
+            root = new AndQueryNode();
+        } else if (selector.get(OR) != null) {
+            clauses = (List<Object>) selector.get(OR);
+            root = new OrQueryNode();
+        }
+
+        //
+        // First handle the simple "field": { "$operator": "value" } clauses.
+        //
+
+        List<Object> basicClauses = new ArrayList<Object>();
+
+        for (Object rawClause: clauses) {
+            Map<String, Object> clause = (Map<String, Object>) rawClause;
+            String field = (String) clause.keySet().toArray()[0];
+            if (!field.startsWith("$")) {
+                basicClauses.add(rawClause);
+            }
+        }
+
+        // Execution step will evaluate each child node and AND or OR the results.
+        for (Object expression: basicClauses) {
+            OperatorExpressionNode node = new OperatorExpressionNode();
+            node.expression = (Map<String, Object>) expression;
+            if (root != null) {
+                root.children.add(node);
+            }
+        }
+
+        //
+        // AND and OR subclauses are handled identically whatever the parent is.
+        // We go through the query twice to order the OR clauses before the AND
+        // clauses, for predictability.
+        //
+
+        // Add subclauses that are OR
+        // TODO - Handle OR subclauses
+
+        // Add subclauses that are AND
+        // TODO - Handle AND subclauses
```

Comment:
Copying response from above:
I believe that the resolution we came to today was to leave this PR as it is wrt the handling of unimplemented functionality and that in the next set of commits and subsequent new PR to account for unimplemented functionality more gracefully there.